sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala

```diff
@@ -331,6 +332,247 @@ case class RoundCeil(child: Expression, scale: Expression)
+      case (ByteType, input: Byte) if (scaleValue <= 0) =>
+        TruncNumber.trunc(input.toLong, scaleValue).toByte
+      case (ShortType, input: Short) if (scaleValue <= 0) =>
+        TruncNumber.trunc(input.toLong, scaleValue).shortValue
+      case (IntegerType, input: Int) if (scaleValue <= 0) =>
+        TruncNumber.trunc(input.toLong, scaleValue).intValue
+      case (LongType, input: Long) if (scaleValue <= 0) =>
+        TruncNumber.trunc(input, scaleValue).longValue
+      case (FloatType, input: Float) =>
+        TruncNumber.trunc(input, scaleValue).floatValue
+      case (DoubleType, input: Double) =>
+        TruncNumber.trunc(input, scaleValue).doubleValue
+      case (DecimalType.Fixed(p, s), input: Decimal) =>
+        Decimal(TruncNumber.trunc(input.toJavaBigDecimal, scaleValue), p, s)
+      case _ => input1
+    }
+  }
+}
+
+object TruncNumber {
+  /**
+   * To truncate whole numbers; byte, short, int, and long types.
+   */
+  def trunc(input: Long, position: Int): Long = {
+    if (position >= 0) {
+      input
+    } else {
+      // Here we truncate the number by the absolute value of the position.
+      // For example, if the input is 123 and the scale is -2, then the result is 100.
+      val pow = Math.pow(10, Math.abs(position)).toLong
+      (input / pow) * pow
+    }
+  }
+
+  /**
+   * To truncate double and float type
+   */
+  def trunc(input: Double, position: Int): BigDecimal = {
+    trunc(jm.BigDecimal.valueOf(input), position)
+  }
+
+  /**
+   * To truncate decimal type
+   */
+  def trunc(input: jm.BigDecimal, position: Int): jm.BigDecimal = {
+    if (input.scale < position) {
+      input
+    } else {
+      val wholePart = input.toBigInteger
+      if (position > 0) {
+        // Here we truncate only the decimal part by the value of the position.
+        val decimalPart = input.remainder(java.math.BigDecimal.ONE)
+        // To avoid overflow during multiplication, we extract the decimal part first,
+        // truncate it and then add the whole part.
+        // For example, if the input is 123.456 and the scale is 2, the result should be 123.45.
+        if (jm.BigDecimal.ZERO.compareTo(decimalPart) == 0) {
+          new jm.BigDecimal(wholePart)
+        } else {
+          val pow = jm.BigDecimal.valueOf(Math.pow(10, position).toLong)
+          val newRemainder = new jm.BigDecimal(decimalPart.multiply(pow).toBigInteger).divide(pow)
+          new jm.BigDecimal(wholePart).add(newRemainder)
+        }
+      } else if (position == 0) {
+        // The position is zero, so we extract the whole part.
+        // For example, if the input is 123.456 and the scale is 0, the result is 123.
+        new jm.BigDecimal(wholePart)
+      } else {
+        // Here we truncate the whole part by the absolute value of the position.
+        // For example, if the input is 123.456 and the scale is -2, the result is 100.
+        if (jm.BigInteger.ZERO.compareTo(wholePart) == 0) {
```

Comment:
in both cases we return `new jm.BigDecimal(wholePart)`. Maybe dedup by combining the conditions:

```
} else if (
  // The position is zero, so we extract the whole part.
  // For example, if the input is 123.456 and the scale is 0, the result is 123.
  position == 0 || 
  // Here we truncate the whole part by the absolute value of the position.
  // For example, if the input is 123.456 and the scale is -2, the result is 100.
  jm.BigInteger.ZERO.compareTo(wholePart) == 0) {
  new jm.BigDecimal(wholePart)
} else {
  val pow = jm.BigInteger.valueOf(Math.pow(10, Math.abs(position)).toLong)
  new jm.BigDecimal(wholePart.divide(pow).multiply(pow), 0)
}
```

Reply:
refactored to avoid code duplication